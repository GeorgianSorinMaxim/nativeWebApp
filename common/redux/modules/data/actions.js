// @flow

export const APP_LAUNCHED = 'data/APP_LAUNCHED';

export const START_FETCH_DATA = 'data/START_FETCH';

export const DATA_FETCH_SUCCESS = 'data/FETCH_SUCCESS';
export const DATA_FETCH_FAILURE = 'data/FETCH_FAILURE';

export function appLaunched() {
  return { type: APP_LAUNCHED };
}

export function startFetchingData() {
  return { type: START_FETCH_DATA };
}

export function dataFetchSuccess(data) {
  return { type: DATA_FETCH_SUCCESS, payload: data };
}

export function dataFetchFailure(error) {
  return { type: DATA_FETCH_FAILURE, payload: error };
}
