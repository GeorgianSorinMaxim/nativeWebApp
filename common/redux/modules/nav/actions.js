// @flow
import { NavigationActions } from 'react-navigation';

const navToPath = path => NavigationActions.navigate({ routeName: path });
// const navToPathWithParams = (path: string, params: Object) =>
//   NavigationActions.navigate({ routeName: path, params });

const navToHome = () => navToPath('Home');
const navToFeature = () => navToPath('Feature');

export const actions = {
  navToHome,
  navToFeature
};
