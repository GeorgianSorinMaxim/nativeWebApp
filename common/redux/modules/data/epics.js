// @flow

import 'rxjs/add/operator/switchMap';
import 'rxjs/add/operator/mapTo';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/pluck';
import 'rxjs/add/observable/of';
import 'rxjs/add/observable/throw';
import 'rxjs/add/observable/fromPromise';
import 'rxjs/add/observable/from';

import { combineEpics } from 'redux-observable';
import type { Epic } from 'redux-observable';
import type { Action } from '../../types';

import * as actions from './actions';

export const startFetchDataEpic: Epic<*, Action, any> = action$ =>
  action$.ofType(actions.APP_LAUNCHED).mapTo(actions.startFetchingData());

export const fetchDataEpic: Epic<*, Action, any> = action$ =>
  action$.ofType(actions.START_FETCH_DATA).switchMap(() =>
    fetch('https://reqres.in/api/users?per_page=12')
      .then(
        response =>
          response.ok
            ? response.json()
            : actions.dataFetchFailure({ error: 'FETCH_ERROR' })
      )
      .then(
        result =>
          result.type === 'REQUEST_FAILED'
            ? result
            : actions.dataFetchSuccess(result)
      )
      .catch(() => actions.dataFetchFailure({ error: 'FETCH_ERROR' }))
  );

export default combineEpics(startFetchDataEpic, fetchDataEpic);
