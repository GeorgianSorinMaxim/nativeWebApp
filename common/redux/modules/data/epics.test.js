import 'rxjs';
import 'isomorphic-fetch';

import { ActionsObservable } from 'redux-observable';

import * as actions from './actions';

import { startFetchDataEpic, fetchDataEpic } from './epics';

const store = {
  dispatch: action => action,
  getState: () => {}
};

describe('Epics', () => {
  describe('startFetchDataEpic', () => {
    it('should call action for start fetching data', () => {
      const action$ = ActionsObservable.of(actions.appLaunched());

      return startFetchDataEpic(action$, store, {})
        .toArray()
        .forEach(result => {
          expect(result).toMatchSnapshot();
        });
    });
  });

  describe('fetchDataEpic', () => {
    it('should call set success action with payload', () => {
      const action$ = ActionsObservable.of(actions.startFetchingData());

      return fetchDataEpic(action$, store, {})
        .toArray()
        .forEach(result => {
          expect(result).toMatchSnapshot();
        });
    });

    it('should call set failure action with error', () => {
      const action$ = ActionsObservable.of(actions.startFetchingData());

      return fetchDataEpic(action$, store, {})
        .toArray()
        .forEach(result => {
          expect(result).toMatchSnapshot();
        });
    });
  });
});
