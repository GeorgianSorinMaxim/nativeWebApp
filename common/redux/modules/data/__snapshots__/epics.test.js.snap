// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Epics fetchDataEpic should call set failure action with error 1`] = `
Array [
  Object {
    "payload": Object {
      "data": Array [
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/calebogden/128.jpg",
          "first_name": "George",
          "id": 1,
          "last_name": "Bluth",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/josephstein/128.jpg",
          "first_name": "Janet",
          "id": 2,
          "last_name": "Weaver",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/olegpogodaev/128.jpg",
          "first_name": "Emma",
          "id": 3,
          "last_name": "Wong",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg",
          "first_name": "Eve",
          "id": 4,
          "last_name": "Holt",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg",
          "first_name": "Charles",
          "id": 5,
          "last_name": "Morris",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg",
          "first_name": "Tracey",
          "id": 6,
          "last_name": "Ramos",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/follettkyle/128.jpg",
          "first_name": "Michael",
          "id": 7,
          "last_name": "Lawson",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/araa3185/128.jpg",
          "first_name": "Lindsay",
          "id": 8,
          "last_name": "Ferguson",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/vivekprvr/128.jpg",
          "first_name": "Tobias",
          "id": 9,
          "last_name": "Funke",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/russoedu/128.jpg",
          "first_name": "Byron",
          "id": 10,
          "last_name": "Fields",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/mrmoiree/128.jpg",
          "first_name": "George",
          "id": 11,
          "last_name": "Edwards",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/hebertialmeida/128.jpg",
          "first_name": "Rachel",
          "id": 12,
          "last_name": "Howell",
        },
      ],
      "page": 1,
      "per_page": 12,
      "total": 12,
      "total_pages": 1,
    },
    "type": "data/FETCH_SUCCESS",
  },
]
`;

exports[`Epics fetchDataEpic should call set success action with payload 1`] = `
Array [
  Object {
    "payload": Object {
      "data": Array [
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/calebogden/128.jpg",
          "first_name": "George",
          "id": 1,
          "last_name": "Bluth",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/josephstein/128.jpg",
          "first_name": "Janet",
          "id": 2,
          "last_name": "Weaver",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/olegpogodaev/128.jpg",
          "first_name": "Emma",
          "id": 3,
          "last_name": "Wong",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/marcoramires/128.jpg",
          "first_name": "Eve",
          "id": 4,
          "last_name": "Holt",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/stephenmoon/128.jpg",
          "first_name": "Charles",
          "id": 5,
          "last_name": "Morris",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/bigmancho/128.jpg",
          "first_name": "Tracey",
          "id": 6,
          "last_name": "Ramos",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/follettkyle/128.jpg",
          "first_name": "Michael",
          "id": 7,
          "last_name": "Lawson",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/araa3185/128.jpg",
          "first_name": "Lindsay",
          "id": 8,
          "last_name": "Ferguson",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/vivekprvr/128.jpg",
          "first_name": "Tobias",
          "id": 9,
          "last_name": "Funke",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/russoedu/128.jpg",
          "first_name": "Byron",
          "id": 10,
          "last_name": "Fields",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/mrmoiree/128.jpg",
          "first_name": "George",
          "id": 11,
          "last_name": "Edwards",
        },
        Object {
          "avatar": "https://s3.amazonaws.com/uifaces/faces/twitter/hebertialmeida/128.jpg",
          "first_name": "Rachel",
          "id": 12,
          "last_name": "Howell",
        },
      ],
      "page": 1,
      "per_page": 12,
      "total": 12,
      "total_pages": 1,
    },
    "type": "data/FETCH_SUCCESS",
  },
]
`;

exports[`Epics startFetchDataEpic should call action for start fetching data 1`] = `
Array [
  Object {
    "type": "data/START_FETCH",
  },
]
`;
